### PR TITLE
Rbd snapshot rollback

### PR DESCRIPTION
**Pull request by GabrielBrascher**
## Description

Add CephSnapshotStrategy to handle RBD revert (rollback) snapshot. In order to support RBD revert (rbd_rollback), this PR adds a CephSnapshotStrategy class to handle Ceph/RBD snapshot actions.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

(i) set `snapshot.backup.to.secondary` to false and restart ACS, ensuring that snapshots are stored directly on primary storage (Ceph/RBD)
(ii) create multiple snapshots with changes (e.g. new files) and stop the VM
(iii) revert to a given snapshot
(iv) assert that rbd_snap_rollback is successfuly executed (remote debug evaulating the code)
(v) start the VM and check that the VM disk content is as expected

### CODE REVIEW

**Comment by GabrielBrascher** [comment]
This PR addresses issue #3498.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@rhtyd rados-java 0.5.0 should be pushed to Maven Central tonight; therefore we need to wait a few hours so we can reckick Travis and blueorangutan. I will ping when rados-java gets updated, thanks!

**Comment by rhtyd** [comment]
Okay @GabrielBrascher ping me when you want regression testing kicked.

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/StorageSystemSnapshotStrategy.java [hunk 1]
```diff
@@ -354,10 +362,9 @@ public boolean revertSnapshot(SnapshotInfo snapshotInfo) {
 
                 s_logger.error(errMsg);
 
-                throw new CloudRuntimeException(errMsg);
+                throw new CloudRuntimeException("Failed to revert a volume to a snapshot state");
```
Comment (wido) [inline engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/StorageSystemSnapshotStrategy.java]: This seems like a very generic message. Can't we add something like the volume name?

Comment (GabrielBrascher) [reply engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/StorageSystemSnapshotStrategy.java]: Good point, I will update this soon. Thanks!

**Comment by wido** [comment]
@rhtyd rados java 0.5.0 is in maven central now

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/CephSnapshotStrategy.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.springframework.stereotype.Component;
+
+import com.cloud.storage.DataStoreRole;
+import com.cloud.storage.Snapshot;
+import com.cloud.storage.Storage.ImageFormat;
+import com.cloud.storage.Storage.StoragePoolType;
+import com.cloud.storage.VolumeVO;
+import com.cloud.storage.dao.VolumeDao;
+
+@Component
```
Comment (rhtyd) [inline engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/CephSnapshotStrategy.java]: We don't have `@Component` scanning, please remove this annotation and instead place the class/bean in respective spring xml context file (if not done already).

File: engine/storage/snapshot/src/main/resources/META-INF/cloudstack/storage/spring-engine-storage-snapshot-storage-context.xml [hunk 3]
```diff
@@ -32,6 +32,9 @@
 
     <bean id="storageSystemSnapshotStrategy"
         class="org.apache.cloudstack.storage.snapshot.StorageSystemSnapshotStrategy" />
+        
+    <bean id="cephSnapshotStrategy"
```
Comment (rhtyd) [inline engine/storage/snapshot/src/main/resources/META-INF/cloudstack/storage/spring-engine-storage-snapshot-storage-context.xml]: +1, just noticed.

**Comment by rhtyd** [comment]
Left a minor remark, tests also LGTM.
We can merge this as soon as any outstanding issues are fixed and we've another lgtm @GabrielBrascher

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/CephSnapshotStrategy.java (hunk 2 above)
Comment (GabrielBrascher) [reply engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/CephSnapshotStrategy.java]: You are right, I will remove the @Component, not needed in this Class. Thanks @rhtyd!

Comment (GabrielBrascher) [reply engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/CephSnapshotStrategy.java]: @rhtyd I have pushed a new commit removing the @Component.

**Pull request merged**